File: sdk/typescript/src/envelope.ts
/**
 * MessageEnvelope build, parse, and serialize.
 */

import { buildMessage, parseMessage, parseMessageAny } from "./message.js";
import type {
  MessageEnvelope,
  MessageEnvelopeHeaders,
  ReplyRoute,
  UnifiedMessage,
} from "./types.js";

/**
 * Wrap UnifiedMessage in a version-1 envelope.
 */
export function buildEnvelope(
  message: UnifiedMessage,
  headers?: MessageEnvelopeHeaders,
): MessageEnvelope {
  return {
    version: "1",
    message,
    ...(headers && Object.keys(headers).length > 0 ? { headers } : {}),
  };
}

/**
 * Build an outbound envelope with text body.
 */
export function buildOutboundEnvelope(params: {
  channel: string;
  accountId: string;
  userId: string;
  text: string;
  agentId?: string;
  replyToMessageId?: string;
  headers?: MessageEnvelopeHeaders;
}): MessageEnvelope {
  const message = buildMessage({
    channel: params.channel,
    accountId: params.accountId,
    userId: params.userId,
    agentId: params.agentId,
    text: params.text,
    replyToMessageId: params.replyToMessageId,
    direction: "outbound",
  });
  return buildEnvelope(message, params.headers);
}

/**
 * Parse MessageEnvelope from JSON string.
 */
export function parseEnvelope(raw: string): MessageEnvelope | null {
  try {
    const obj = JSON.parse(raw) as Record<string, unknown>;
    if (!obj || typeof obj !== "object") return null;

    if (obj.version === "1" && obj.message && typeof obj.message === "object") {
      const msg = obj.message as UnifiedMessage;
      if (!msg.messageId || !msg.source?.channel) return null;
      return {
        version: "1",
        message: msg,
        headers: (obj.headers as MessageEnvelopeHeaders) ?? undefined,
      };
    }

    const legacy = parseMessage(raw);
    if (legacy) {
      return { version: "1", message: legacy };
    }
    return null;
  } catch {
    return null;
  }
}

/**
 * Parse MessageEnvelope from string, Buffer, Uint8Array, or object.
 */
export function parseEnvelopeAny(
  input: string | Buffer | Uint8Array | unknown,
): MessageEnvelope | null {
  if (typeof input === "string") return parseEnvelope(input);
  if (typeof Buffer !== "undefined" && Buffer.isBuffer(input)) {
    return parseEnvelope(input.toString("utf-8"));
  }
  if (input instanceof Uint8Array) {
    return parseEnvelope(new TextDecoder().decode(input));
  }
  if (typeof input === "object" && input !== null) {
    const o = input as Record<string, unknown>;
    if (o.version === "1" && o.message) {
      return o as unknown as MessageEnvelope;
    }
    const unified = parseMessageAny(input);
    if (unified) {
      return { version: "1", message: unified };
    }
  }
  return null;
}

/**
 * Extract correlation/idempotency from message metadata.
 */
export function extractRoutingMetadata(msg: UnifiedMessage): {
  correlationId?: string;
  idempotencyKey?: string;
} {
  const meta = msg.metadata ?? {};
  return {
    correlationId: typeof meta.correlationId === "string" ? meta.correlationId : undefined,
    idempotencyKey: typeof meta.idempotencyKey === "string" ? meta.idempotencyKey : undefined,
  };
}

/**
 * Merge replyRoute into envelope headers.
 */
export function mergeReplyRouteIntoHeaders(
  headers: MessageEnvelopeHeaders | undefined,
  replyRoute: ReplyRoute | undefined,
): MessageEnvelopeHeaders | undefined {
  if (!replyRoute || Object.keys(replyRoute).length === 0) {
    return headers;
  }
  return { ...headers, replyRoute };
}

/**
 * Read reply route from envelope headers.
 */
export function getReplyRoute(envelope: MessageEnvelope): ReplyRoute | undefined {
  return envelope.headers?.replyRoute;
}

/**
 * Serialize envelope to JSON string.
 */
export function serializeEnvelope(envelope: MessageEnvelope): string {
  return JSON.stringify(envelope);
}

File: sdk/python/openclaw_message_sdk/envelope.py
"""MessageEnvelope build, parse, and serialize."""

from __future__ import annotations

import json
from typing import Any

from openclaw_message_sdk.message import build_message, parse_message, parse_message_any
from openclaw_message_sdk.types import (
    MessageEnvelope,
    MessageEnvelopeHeaders,
    ReplyRoute,
    UnifiedMessage,
)


def build_envelope(
    message: UnifiedMessage,
    headers: MessageEnvelopeHeaders | None = None,
) -> MessageEnvelope:
    """Wrap UnifiedMessage in a version-1 envelope."""
    envelope: MessageEnvelope = {"version": "1", "message": message}
    if headers:
        envelope["headers"] = headers
    return envelope


def build_outbound_envelope(
    *,
    channel: str,
    account_id: str,
    user_id: str,
    text: str,
    agent_id: str | None = None,
    reply_to_message_id: str | None = None,
    headers: MessageEnvelopeHeaders | None = None,
) -> MessageEnvelope:
    """Build an outbound envelope with text body."""
    params: dict[str, Any] = {
        "channel": channel,
        "accountId": account_id,
        "userId": user_id,
        "text": text,
        "direction": "outbound",
    }
    if agent_id:
        params["agentId"] = agent_id
    if reply_to_message_id:
        params["replyToMessageId"] = reply_to_message_id
    message = build_message(params)  # type: ignore[arg-type]
    return build_envelope(message, headers)


def parse_envelope(raw: str) -> MessageEnvelope | None:
    """Parse MessageEnvelope from JSON string."""
    try:
        obj = json.loads(raw)
        if not isinstance(obj, dict):
            return None
        if obj.get("version") == "1" and isinstance(obj.get("message"), dict):
            msg = obj["message"]
            source = msg.get("source") if isinstance(msg, dict) else None
            if not isinstance(msg.get("messageId"), str) or not isinstance(source, dict):
                return None
            if not isinstance(source.get("channel"), str):
                return None
            return obj  # type: ignore[return-value]
        legacy = parse_message(raw)
        if legacy:
            return {"version": "1", "message": legacy}
        return None
    except json.JSONDecodeError:
        return None


def parse_envelope_any(raw: str | bytes | bytearray | dict[str, Any]) -> MessageEnvelope | None:
    """Parse MessageEnvelope from string, bytes, or dict."""
    if isinstance(raw, (bytes, bytearray)):
        return parse_envelope(raw.decode("utf-8"))
    if isinstance(raw, str):
        return parse_envelope(raw)
    if isinstance(raw, dict):
        if raw.get("version") == "1" and raw.get("message"):
            return raw  # type: ignore[return-value]
        unified = parse_message_any(raw)
        if unified:
            return {"version": "1", "message": unified}
    return None


def get_reply_route(envelope: MessageEnvelope) -> ReplyRoute | None:
    """Read reply route from envelope headers."""
    headers = envelope.get("headers")
    if not headers:
        return None
    route = headers.get("replyRoute")
    return route if isinstance(route, dict) else None


def serialize_envelope(envelope: MessageEnvelope) -> str:
    """Serialize envelope to JSON string."""
    return json.dumps(envelope, ensure_ascii=False)
